ex_4/hashizume/EX4.py

```diff
@@ -0,0 +1,254 @@
+
+    Parameters
+    ----------
+    x : input
+    dBref : decibel reference
+    """
+
+    y = 20 * np.log10(x / dBref)
+    return y
+
+
+def lin(x, dBref):
+    """
+    db → linear conversion
+
+    Parameters
+    ----------
+    x : input
+    dBref : decibel reference
+    """
+
+    y = dBref * 10 ** (x / 20)
+    return y
+
+
+def norm(spectrum, dBref):
+    """
+    normalization
+
+    Parameters
+    ----------
+    x : input
+    dBref : decibel reference
+    """
+
+    spectrum = lin(spectrum, dBref)
+    spectrum = np.abs(spectrum / (len(spectrum) / 2))
+    spectrum = db(spectrum, dBref)
+    return spectrum
+
+
+def levinson(r):
+    """
+    levinson durbin algorithm
+    """
+    a = np.zeros(len(r))
+    a[0] = 1
+    sigma = r[0]
+    for p in range(1, a.shape[0]):
+        w = np.sum(a[:p] * r[p:0:-1], axis=0)
+        k = w / sigma
+        sigma = sigma - k * w
+
+        a[1 : p + 1] = a[1 : p + 1] - k * a[p - 1 :: -1]
+    e = np.sqrt(sigma)
+    return a, e
+
+
+def lpc(x, win_length, deg):
+    ac = AutoCorrelation(x, x.shape[0])
+    r = ac[:deg]
+    a, e = levinson(r)
+    env = e / np.fft.rfft(a, win_length, axis=0)
+    env = librosa.amplitude_to_db(np.abs(env))
+
+    return env
+
+
+def main():
+    args = sys.argv
+    wav_filename = args[1]
+    origin_signal, sr = librosa.load(wav_filename)
+
+    # F0 estimate
+    # Autocorrelation
+    win_length = 1024
+    hop_length = 512
+
+    f0series = calc_AC(origin_signal, win_length, hop_length, sr)
+
+    f0series = np.concatenate([[f0series[0]], f0series, [f0series[-1]]])
+    s = np.array(origin_signal).astype(np.float32)
+
+    spec = librosa.stft(s, win_length, hop_length)
+    spec_db = librosa.amplitude_to_db(np.abs(spec))
+    librosa.display.specshow(spec_db, y_axis="log")
+
+    plt.plot(np.arange(spec_db.shape[1]), f0series, color="b", linewidth=2.0)
+    plt.xlabel("time[s]")
+    plt.ylabel("freqency[Hz]")
+    plt.title("AC")
+    plt.savefig("AC")
+    plt.clf()
+    plt.close
+
+    # cepstrum
+
+    # fft
+    spec = np.fft.fft(origin_signal)
+    # linear → db
+    spec_db = db(spec, 2e-5)
+    # ifft
+    ceps_db = np.real(np.fft.ifft(spec_db))
+    # LPF
+    index = 50
+    ceps_db[index : len(ceps_db) - index] = 0
+    # fft ceps_db_low : spectrum envelope
+    ceps_db_low = np.fft.fft(ceps_db)
+
+    # calculate the amplitude component of the audio spectrum
+    spec_db_amp = norm(spec_db, 2e-5)
+    # calculate the amplitude component of the spectrum envelope
+    ceps_db_low_amp = norm(ceps_db_low, 2e-5)
+
+    # LPC
+    env_LPC = lpc(origin_signal, win_length, 32)
+
+    # plot
+
+    # axis settting
+    frequency = np.linspace(0, sr, len(origin_signal))
+    quefrency = np.arange(0, len(origin_signal)) / sr
+
+    plt.xlabel("Frequency [Hz]")
+    plt.ylabel("SPL [dB]")
+
+    plt.title("envelope")
+
+    plt.plot(frequency, spec_db_amp)
+    plt.plot(frequency, ceps_db_low_amp, lw="4")
+    plt.plot(env_LPC)
```

Comment:
```suggestion
    plt.plot(frequency, spec_db_amp)
    plt.plot(frequency, ceps_db_low_amp, lw="4")
    plt.plot(frequency, env_LPC)
```
サイズが合えば lpc の方にも渡せます.